README.md

```diff
@@ -1,16 +1,1263 @@
+
+### 2.2.3 Batch
+
+A [Batch](#5-batch) is a way of requesting more than one action at once.
+
+### 2.2.4 Invocation
+
+An [Invocation](#6-invocation) is the cryptographically signed container for a Batch. This is the step that "forces" the "deferred" Closure.
+
+### 2.2.5 Pointers
+
+An [Invocation Pointer](#7-pointer) identifies a specific invocation. An Invoked Task Pointer points to a unique Task inside an Invocation.
+
+### 2.2.6 Result
+
+A [Result](#8-result) is the output of a Closure.
+
+### 2.2.7 Receipt
+
+A [Receipt](#9-receipt) describes the output of an invocation, referenced by its Invocation Pointer.
+
+### 2.2.8 Promise
+
+A [promise](#10-promise) is a reference to the receipt of an action that has yet to return a receipt.
+
+## 2.3 IPLD Schema
+
+``` ipldsch
+type Closure struct {
+  with   URI
+  do     Ability
+  inputs Any
+}
+
+type Task struct {
+  with   URI
+  do     Ability
+  inputs Any
+  meta {String : Any} (implicit {})
+}
+
+type Batch union {
+  | Named {String : Task}
+  | List  [Task]
+}
+
+type Invocation struct {
+  uiv  SemVer
+  run  Batch
+  prf  [&UCAN]
+  nnc  String
+  meta {String : Any} (implicit {})
+  sig  Varsig
+}
+
+type InvocationPointer union {
+  | "/" -- Relative to the current invocation
+  | &Invocation
+}
+
+type InvokedTaskPointer struct {
+  envl  InvocationPointer
+  label String
+} representation tuple
+
+type Receipt struct {
+  ran  &InvocationPointer
+  out  {String : Result}
+  rec  {String : &Receipt}
+  meta {String : Any}
+  sig  Varsig
+} 
+
+type Result union {
+  | Any ("ok")  -- Success
+  | Any ("err") -- Failure
+} representation keyed
+
+type Promise union {
+  | InvokedTaskPointer "promise/ok"
+  | InvokedTaskPointer "promise/err"
+  | InvokedTaskPointer "promise/*"
+} representation keyed
+```
+
+# 3 Closure
+
+A Closure is the smallest unit of work that can be requested from a UCAN. It describes one `(resource, ability, inputs)` triple. The `inputs` field is free form, and depend on the specific resource and ability being interacted with, and not described in this specification.
+
+Using the JavaScript analogy from the introduction, a Closure is similar to wrapping a call in an anonymous function:
+
+``` json
+{
+  "with": "mailto://alice@example.com",
+  "do": "msg/send",
+  "inputs": {
+    "to": ["bob@example.com", "carol@example.com"],
+    "subject": "hello",
+    "body": "world"
+  }
+}
+```
+
+``` js
+// Pseudocode
+() => msg.send("mailto:alice@example.com", {
+  to: ["bob@example.com", "carol@example.com"],
+  subject: "hello",
+  body: "world"
+})
+```
+
+Later, when we explore [Promises](# 10-promise), this also includes capturing the promise:
+
+``` json
+{
+  "mailingList": {
+    "with": "https://example.com/mailinglist",
+    "do": "crud/read"
+  },
+  "sendEmail": {
+    "with": "mailto://alice@example.com",
+    "do": "msg/send",
+    "inputs": {
+      "to": {"promise/*": ["/", "get-mailing-list"]}
```

Comment:
```suggestion
      "to": {"promise/*": ["/", "mailingList"]}
```

I'm not sure if this is the best replacement. Where did `get-mailing-list` get defined?

Reply:
Ah, indeed, this is actually probably the other way around to keep it consistent. A lot of these are me updating various versions of the spec as it was evolving quickly; thanks for the `suggestion` @gobengo !